### ISSUE TITLE
poly: investigate ConvertToLLVMPatternInterface

### ISSUE THREAD

**Issue by j2kun** [google/heir#103]
From [review on ](https://github.com/google/heir/pull/94#issuecomment-1682621423)

> Reading the [MLIR newsletter](https://discourse.llvm.org/t/mlir-news-53th-edition-16th-august-2023/72805) made me aware of ConvertToLLVMPatternInterface and I'd recommend we add it to the poly dialect :)

As I understand it, this interface is already implemented for math, arith, func, complex, cf, index, and memref, so I'm not entirely sure what benefit we'd get from implementing it vs lowering to those dialects. But it's worth learning more about this interface to see how appropriate it is for poly.

**Comment by AlexanderViand-Intel** [comment]
I understood [this description](https://github.com/llvm/llvm-project/commit/4529797a9d7c19105815cc9a3f19571b5fca2d06) to mean that something like our poly->various->llvm approach might still be in scope, to quote the commit message: "Since these conversions can involve intermediate dialects [...]"

**Comment by asraa** [comment]
Right, I think from what I see here if we implement this interface, then we would be able to use `--convert-to-llvm` pass, which will scan for dialects that implement this pattern interface. There's one point around bufferizing step that I don't understand. If we did the following:

(1) Updating the `PolyToStandard` conversion code to implement `ConvertToLLVMPatternInterface`. With regard to the intermediate dialect transformations, I think this will all be automatic so long as our intermediate dialects also register their `ConvertToLLVMPatternInterface`. This is what I don't quite understand: we need to bufferize linalg at some point, before lowering linalg to LLVM through std/memref/etc. I don't know whether it's feasible to implement this pattern before bufferizing.

(2) Registering that `Poly` implements the `ConvertToLLVMPatternInterface` 
```
  registry.addExtension(+[](MLIRContext *ctx, heir::PolyDialect *dialect) {
    dialect->addInterfaces<PolyToLLVMDialectInterface>();
  });
``` 
This seems easy.

**Comment by j2kun** [comment]
I'm hoping we could just reuse `arith::populateArithToLLVMConversionPatterns` and similar when implementing that interface (only having to add in our poly-to-tensor patterns or what have you), and we may have to extend `LLVMTypeConverter`...

**Comment by j2kun** [comment]
This should move to upstream MLIR once we upstream PolynomialToStandard.

**Comment by j2kun** [comment]
Looking around more upstream, there's no high-level dialects that implement this interface, just ones that go directly to LLVM. So closing.
